### PR TITLE
ValidatorSet: Seperate preimages from enrollments

### LINKED ISSUE #2376: Preimage gossip loop
```
Aug 10 07:31:16 eu-002 docker[652000]: 2021-08-10 07:31:16,269 Info [boa1xzval3ah8z7ewhuzx6mywveyr79f24w49rdypwgurhjkr8z2ke2mycftv9n] - Accepted preimage: { utxo: 0x9b27...6b2c, hash: 0xdc20...6817, height: { value: 5778 } }
Aug 10 07:31:16 eu-002 docker[652000]: 2021-08-10 07:31:16,275 Info [boa1xzval3ah8z7ewhuzx6mywveyr79f24w49rdypwgurhjkr8z2ke2mycftv9n] - Accepted preimage: { utxo: 0x210f...3b64, hash: 0xa90a...1c88, height: { value: 5777 } }
Aug 10 07:31:16 eu-002 docker[652000]: 2021-08-10 07:31:16,280 Info [boa1xzval3ah8z7ewhuzx6mywveyr79f24w49rdypwgurhjkr8z2ke2mycftv9n] - Accepted preimage: { utxo: 0x9b27...6b2c, hash: 0xdc20...6817, height: { value: 5778 } }
Aug 10 07:31:16 eu-002 docker[652000]: 2021-08-10 07:31:16,285 Info [boa1xzval3ah8z7ewhuzx6mywveyr79f24w49rdypwgurhjkr8z2ke2mycftv9n] - Accepted preimage: { utxo: 0xab19...1255, hash: 0x4f55...3758, height: { value: 5778 } }
Aug 10 07:31:16 eu-002 docker[652000]: 2021-08-10 07:31:16,290 Info [boa1xzval3ah8z7ewhuzx6mywveyr79f24w49rdypwgurhjkr8z2ke2mycftv9n] - Accepted preimage: { utxo: 0xdb76...2a0a, hash: 0x95f8...2c9d, height: { value: 5778 } }
Aug 10 07:31:16 eu-002 docker[652000]: 2021-08-10 07:31:16,295 Info [boa1xzval3ah8z7ewhuzx6mywveyr79f24w49rdypwgurhjkr8z2ke2mycftv9n] - Accepted preimage: { utxo: 0xdb76...2a0a, hash: 0x95f8...2c9d, height: { value: 5778 } }
Aug 10 07:31:16 eu-002 docker[652000]: 2021-08-10 07:31:16,301 Info [boa1xzval3ah8z7ewhuzx6mywveyr79f24w49rdypwgurhjkr8z2ke2mycftv9n] - Accepted preimage: { utxo: 0x210f...3b64, hash: 0xa90a...1c88, height: { value: 5777 } }
Aug 10 07:31:16 eu-002 docker[652000]: 2021-08-10 07:31:16,306 Info [boa1xzval3ah8z7ewhuzx6mywveyr79f24w49rdypwgurhjkr8z2ke2mycftv9n] - Accepted preimage: { utxo: 0x9b27...6b2c, hash: 0xdc20...6817, height: { value: 5778 } }
Aug 10 07:31:16 eu-002 docker[652000]: 2021-08-10 07:31:16,312 Info [boa1xzval3ah8z7ewhuzx6mywveyr79f24w49rdypwgurhjkr8z2ke2mycftv9n] - Accepted preimage: { utxo: 0xab19...1255, hash: 0x4f55...3758, height: { value: 5778 } }
Aug 10 07:31:16 eu-002 docker[652000]: 2021-08-10 07:31:16,317 Info [boa1xzval3ah8z7ewhuzx6mywveyr79f24w49rdypwgurhjkr8z2ke2mycftv9n] - Accepted preimage: { utxo: 0xab19...1255, hash: 0x4f55...3758, height: { value: 5778 } }
Aug 10 07:31:16 eu-002 docker[652000]: 2021-08-10 07:31:16,323 Info [boa1xzval3ah8z7ewhuzx6mywveyr79f24w49rdypwgurhjkr8z2ke2mycftv9n] - Accepted preimage: { utxo: 0xab19...1255, hash: 0x4f55...3758, height: { value: 5778 } }
Aug 10 07:31:16 eu-002 docker[652000]: 2021-08-10 07:31:16,333 Info [boa1xzval3ah8z7ewhuzx6mywveyr79f24w49rdypwgurhjkr8z2ke2mycftv9n] - Accepted preimage: { utxo: 0xab19...1255, hash: 0x4f55...3758, height: { value: 5778 } }
Aug 10 07:31:16 eu-002 docker[652000]: 2021-08-10 07:31:16,383 Info [boa1xzval3ah8z7ewhuzx6mywveyr79f24w49rdypwgurhjkr8z2ke2mycftv9n] - Accepted preimage: { utxo: 0xab19...1255, hash: 0x4f55...3758, height: { value: 5778 } }
Aug 10 07:31:16 eu-002 docker[652000]: 2021-08-10 07:31:16,389 Info [boa1xzval3ah8z7ewhuzx6mywveyr79f24w49rdypwgurhjkr8z2ke2mycftv9n] - Accepted preimage: { utxo: 0xab19...1255, hash: 0x4f55...3758, height: { value: 5778 } }
Aug 10 07:31:16 eu-002 docker[652000]: 2021-08-10 07:31:16,395 Info [boa1xzval3ah8z7ewhuzx6mywveyr79f24w49rdypwgurhjkr8z2ke2mycftv9n] - Accepted preimage: { utxo: 0xab19...1255, hash: 0x4f55...3758, height: { value: 5778 } }
Aug 10 07:31:16 eu-002 docker[652000]: 2021-08-10 07:31:16,400 Info [boa1xzval3ah8z7ewhuzx6mywveyr79f24w49rdypwgurhjkr8z2ke2mycftv9n] - Accepted preimage: { utxo: 0xab19...1255, hash: 0x4f55...3758, height: { value: 5778 } }
Aug 10 07:31:16 eu-002 docker[652000]: 2021-08-10 07:31:16,405 Info [boa1xzval3ah8z7ewhuzx6mywveyr79f24w49rdypwgurhjkr8z2ke2mycftv9n] - Accepted preimage: { utxo: 0x210f...3b64, hash: 0xa90a...1c88, height: { value: 5777 } }
Aug 10 07:31:16 eu-002 docker[652000]: 2021-08-10 07:31:16,411 Info [boa1xzval3ah8z7ewhuzx6mywveyr79f24w49rdypwgurhjkr8z2ke2mycftv9n] - Accepted preimage: { utxo: 0xdb76...2a0a, hash: 0x95f8...2c9d, height: { value: 5778 } }
````

Currently two of live env nodes keep accepting same preimages, which should not be possible. this probably creates an endless gossip loop. This seems to be happenning on eu-002 only right now.

### PR DESCRIPTION
**Pull request by omerfirmak**
Fixes #2376

### CODE REVIEW

**Comment by Geod24** [comment]
Hum, could we not fail on them ? Since they are tied to an UTXO, updating it would be better, wouldn't it ?

**Comment by omerfirmak** [comment]
> 
> 
> Hum, could we not fail on them ? Since they are tied to an UTXO, updating it would be better, wouldn't it ?

We would need to create an enrollment out of thin air for that, without seeing it in a block.

**Comment by Geod24** [comment]
Would we ? We could just separate the informations about enrollments and preimages. Preimages aren't tied to enrollments, they are tied to UTXO.

**Comment by omerfirmak** [comment]
> 
> 
> Would we ? We could just separate the informations about enrollments and preimages. Preimages aren't tied to enrollments, they are tied to UTXO.

But they can only be validated with the previous preimage and that does not exist if that UTXO was never used for an enrollment.

**Comment by Geod24** [comment]
> But they can only be validated with the previous preimage and that does not exist if that UTXO was never used for an enrollment.

Yeah but that's already detected. The error we are seeing happens when a node rejects preimages that are past the enrollment cycle.

**Comment by omerfirmak** [comment]
> 
> 
> > But they can only be validated with the previous preimage and that does not exist if that UTXO was never used for an enrollment.
> 
> Yeah but that's already detected. The error we are seeing happens when a node rejects preimages that are past the enrollment cycle.

Updated to reflect that preimages are only tied to UTXOs.

File: source/agora/consensus/state/ValidatorSet.d [hunk 1]
```diff
@@ -323,15 +332,14 @@ public class ValidatorSet
...
+            auto utxo = Hash(row.peek!(const(char)[], PeekMode.slice)(2));
             ret ~= ValidatorInfo(
                 /* enrolled: */ Height(row.peek!(ulong)(0)),
                 /* address:  */ PublicKey.fromString(row.peek!(char[], PeekMode.slice)(1)),
-                /* preimage: */ PreImageInfo(
-                /*     utxo:   */ Hash(row.peek!(const(char)[], PeekMode.slice)(2)),
-                /*     hash:   */ Hash(row.peek!(const(char)[], PeekMode.slice)(3)),
-                /*     height: */ Height(row.peek!(ulong)(4)),
-                    ),
+                /* preimage: */ this.getPreimage(utxo),
```
Comment (Geod24) [inline source/agora/consensus/state/ValidatorSet.d]: I was hoping for a JOIN ? :)

File: source/agora/consensus/state/ValidatorSet.d [hunk 2]
```diff
@@ -93,6 +93,11 @@ public class ValidatorSet
             "cycle_length INTEGER, enrolled_height INTEGER, " ~
             "height INTEGER, preimage TEXT, nonce TEXT, slashed_height INTEGER,
```
Comment (Geod24) [inline source/agora/consensus/state/ValidatorSet.d]: This needs updating

File: source/agora/consensus/state/ValidatorSet.d [hunk 3]
```diff
@@ -93,6 +93,11 @@ public class ValidatorSet
...
             PRIMARY KEY (key, enrolled_height))");
+
+        // create the table for preimages if it doesn't exist yet
+        this.db.execute("CREATE TABLE IF NOT EXISTS preimages " ~
+            "(key TEXT, height INTEGER, preimage TEXT,
+            PRIMARY KEY (key))");
```
Comment (Geod24) [inline source/agora/consensus/state/ValidatorSet.d:100]: https://github.com/bosagora/agora/blob/v0.x.x/doc/Database.md

**Pull request merged**